### PR TITLE
[PyTorch] Make c10::irange(x) generate the same assembly as for loop

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86841

`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

### CODE REVIEW

File: c10/test/util/irange_test.cpp [hunk 1]
```diff
@@ -6,6 +6,20 @@
 
 using namespace ::testing;
 
+void f(int x);
+
+void iterateOldStyle(int bound) {
```
Comment (r-barnes) [inline c10/test/util/irange_test.cpp:11]: What is this being used for?

File: c10/test/util/irange_test.cpp [hunk 2]
```diff
@@ -6,6 +6,20 @@
...
+  for (int ii = 0; ii < bound; ++ii) {
+    f(ii);
+  }
+}
+
+void iterateIrange(int bound) {
```
Comment (r-barnes) [inline c10/test/util/irange_test.cpp:17]: Is this just to test compilation?

File: c10/util/irange.h [hunk 3]
```diff
@@ -54,20 +63,22 @@ struct integer_iterator : std::iterator<std::input_iterator_tag, I> {
 
 template <
     typename I,
+    bool one_sided = false,
```
Comment (r-barnes) [inline c10/util/irange.h:66]: I don't see a place where `one_sided` gets set to true?

**Comment by r-barnes** [comment]
@swolchok - Thanks for this. Since `if constexpr` is part of C++17, we need to hold off on this until PyTorch completes its migration to C++17.

File: c10/test/util/irange_test.cpp (hunk 2 above)
Comment (swolchok) [reply c10/test/util/irange_test.cpp:17]: yep, can remove if you like

File: c10/util/irange.h [hunk 4]
```diff
@@ -95,11 +106,8 @@ template <
     typename Integer,
     typename std::enable_if<std::is_integral<Integer>::value, bool>::type =
         true>
-integer_range<Integer> irange(Integer end) {
-  // If end<=begin then the range is empty; we can achieve this effect by
-  // choosing the larger of {0, end} as the loop terminator
-  // Handles the case where end<0. irange only works for ranges >=0
-  return {Integer(), std::max(Integer(), end)};
+integer_range<Integer, true> irange(Integer end) {
```
Comment (swolchok) [inline c10/util/irange.h:109]: here is where one_sided gets set to true

**Comment by swolchok** [comment]
#85969 is the PR we are blocked on here.

File: c10/util/irange.h [hunk 5]
```diff
@@ -39,11 +40,19 @@ struct integer_iterator : std::iterator<std::input_iterator_tag, I> {
   }
 
   bool operator==(const integer_iterator& other) const {
-    return value == other.value;
+    if constexpr (one_sided) {
```
Comment (swolchok) [inline c10/util/irange.h:43]: I think we can just make this an ordinary `if` statement and proceed.

File: aten/src/ATen/native/TensorShape.cpp [hunk 6]
```diff
@@ -2889,7 +2889,7 @@ Tensor squeeze_qtensor(const Tensor& self, c10::optional<int64_t> dim) {
     const auto* per_channel_quantizer = static_cast<at::PerChannelAffineQuantizer*>(quantizer.get());
     auto axis = per_channel_quantizer->axis();
     int64_t shift = 0;
-    integer_range<int64_t> dims = dim.has_value() ? integer_range<int64_t>{dim.value(), dim.value() + 1} : c10::irange(self.dim());
+    integer_range<int64_t> dims = dim.has_value() ? integer_range<int64_t>{dim.value(), dim.value() + 1} : c10::irange(0, self.dim());
```
Comment (r-barnes) [inline aten/src/ATen/native/TensorShape.cpp:2892]: Is this a test artefact?

**Review by r-barnes — APPROVED** [review]

File: aten/src/ATen/native/TensorShape.cpp (hunk 6 above)
Comment (swolchok) [reply aten/src/ATen/native/TensorShape.cpp:2892]: No, it's necessary for the two forks of the ?: operator to have the same type.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
If you believe that failures are unrelated to the PR, feel free to merge it with `-g`.. But enabling C++17 across all our platforms feels a bit convoluted, but hopefully will get it done soon.

**Comment by swolchok** [comment]
failure definitely looks unrelated

**Pull request closed without merging**